src/runtime-rs/src/shim/src/event_publisher.rs

```diff
@@ -0,0 +1,220 @@
+use std::process::{Command, Stdio};
+use std::sync::mpsc::{channel, Receiver, Sender};
+use std::sync::{Arc, Mutex};
+
+use anyhow::{Context, Result};
+use log::{debug, error, info, warn};
+use protobuf::{well_known_types::Any, Message};
+
+use agent::Agent;
+use shim_proto::events_task::TaskOOM;
+
+const TASK_OOMEVENT_TOPIC: &str = "/tasks/oom";
+
+pub trait Event {
+    fn r#type(&self) -> String;
+    fn type_url(&self) -> String;
+}
+
+impl Event for TaskOOM {
+    fn r#type(&self) -> String {
+        TASK_OOMEVENT_TOPIC.to_string()
+    }
+
+    fn type_url(&self) -> String {
+        "containerd.events.TaskOOM".to_string()
+    }
+}
+
+pub struct EventData {
+    r#type: String,
+    type_url: String,
+    value: Vec<u8>,
+}
+
+impl EventData {
+    pub fn new<T: Event + Message>(event: &T) -> Result<Self> {
+        let value = event
+            .write_to_bytes()
+            .context(format!("write to data for event {}", &event.r#type()))?;
+
+        Ok(Self {
+            r#type: event.r#type(),
+            type_url: event.type_url(),
+            value,
+        })
+    }
+}
+
+/// Event publisher to send events to the Containerd.
+///
+/// It will automatically forward all TaskOOM events from agent to Containerd.
+/// TODO: support grace shutdown.
+pub struct EventPublisher {
+    containerd_binary: String,
+    address: String,
+    namespace: String,
+    sender: Option<Sender<EventData>>,
+}
+
+impl EventPublisher {
+    /// Create a new instance of ['EventPublisher'].
+    pub fn new(containerd_binary: &str, address: &str, namespace: &str) -> Self {
+        Self {
+            containerd_binary: containerd_binary.to_string(),
+            address: address.to_string(),
+            namespace: namespace.to_string(),
+            sender: None,
+        }
+    }
+
+    /// Start to forward events from
+    pub fn start(&mut self, agent: Arc<dyn Agent>, event_publisher: Arc<Mutex<EventPublisher>>) {
+        // start run forwarder
+        let (sender, receiver) = channel();
+
+        self.sender = Some(sender);
+        self.run_forwarder(receiver);
+        self.start_oom_watcher(agent, event_publisher);
+    }
+
+    /// Send a event to the forwarder.
+    pub fn send_event<T: Event + Message>(&self, event: &T) -> Result<()> {
+        let event_data = EventData::new(event).context("new event data")?;
+
+        if let Some(sender) = self.sender.clone() {
+            sender.send(event_data).context("send event")?;
+        }
+
+        Ok(())
+    }
+
+    /// Send a OOM event for `container_id` to the forwarder.
+    pub fn send_oom_event(&self, container_id: &str) {
+        info!("send oom event for container {}", container_id);
+        let e = TaskOOM {
+            container_id: container_id.to_string(),
+            ..Default::default()
+        };
+
+        if let Err(err) = self.send_event(&e) {
+            error!(
+                "failed to send oom event for {} error {:?}",
+                container_id, err
```

Comment:
Ideally, all the new code here would use structured logging. Something like the following (untested):

```rust
error!("failed to send oom event"; "container-id" => container_id, "error" => format!("{:?}", err));
```


See: https://github.com/kata-containers/tests/tree/main/cmd/log-parser

Reply:
Leave it to the later revisions. May change log libs to https://github.com/kata-containers/kata-containers/tree/main/src/libs/logging

Reply:
Fixed in the second revision: now use https://github.com/kata-containers/kata-containers/tree/main/src/libs/logging as the log libs to generate structured logging